sshd-core/src/main/java/org/apache/sshd/client/SshClient.java

```diff
@@ -586,9 +587,14 @@ protected ConnectFuture doConnect(
                                         if (f5.isConnected()) {
                                             ClientSession clientSession = f5.getClientSession();
                                             clientSession.setAttribute(TARGET_SERVER, address);
-                                            connectFuture.setSession(clientSession);
-                                            proxySession.addCloseFutureListener(f6 -> clientSession.close(true));
-                                            clientSession.addCloseFutureListener(f6 -> proxySession.close(true));
+                                            try {
+                                                setSessionSafely(connectFuture, clientSession);
+                                                proxySession.addCloseFutureListener(f6 -> clientSession.close(true));
+                                                clientSession.addCloseFutureListener(f6 -> proxySession.close(true));
```

Comment:
This isn't correct yet. If there is a timeout, the proxySession also needs to be closed.